Tutorial 11: Advanced Account Updates (#348)
o1-labs/docs2: docs/zkapps/tutorials/11-advanced-account-updates.mdx

---
title: 'Tutorial 11: Advanced Account Updates'
hide_title: true
sidebar_label: 'Tutorial 11: Advanced Account Updates'
---

:::info

Please note that zkApp programmability is not yet available on Mina Mainnet, but
zkApps can now be deployed to Berkeley Testnet.

:::

:::note

This tutorial was last tested with [SnarkyJS](https://www.npmjs.com/package/snarkyjs) 0.9.6.

:::

# Tutorial 11: Advanced Account Updates

## zkAppManager Accounts and Tokens

## Overview:

In the last tutorial, you learned how zkApp transactions are composed of AccountUpdates and about the structure of Account Updates.

In this tutorial, you learn about more advanced features of account updates. Namely tokens, zkApp Manager Accounts, and account update assertions.

An earlier tutorial covered how each token is associated with a manager account and how the manager account determines rules for token minting, burning, and transfer.

> Under the hood, what is happening is actually a bit more general than that. The Manager Account for a token actually controls all properties of token accounts.

Reviewer: ```suggestion
Internally, what is happening is actually a bit more general than that. The Manager Account for a token actually controls all properties of token accounts.
```

> ## zkApp Manager Accounts & Tokens

Reviewer: ```suggestion
## zkApp Manager Accounts and Tokens
```

For an Account Update to be able to modify state (including sending tokens) on a token account, two things must be true:

> 1. The account update must meet the criteria of the permissions on the zkApp account itself. If the zkApp Account is setup with the `proof` permission, this means the Account Update proof must pass the verification key in the account

Reviewer: ```suggestion
1. The account update must meet the criteria of the permissions on the zkApp account itself. If the zkApp Account is set up with the `proof` permission, the Account Update proof must pass the verification key in the account.
```

> 2. The account update must set `mayUseToken` to True. An account update on a token account cannot set this property to true itself however. It must get its account update “approved” by its corresponding manager account. The manager account can approve accounts based on its `access` permission.

Reviewer: ```suggestion
2. The account update must set `mayUseToken` to True. However, an account update on a token account cannot set this property to true. It must get its account update “approved” by its corresponding manager account. The manager account can approve accounts based on its `access` permission.
```

## zkApp Token Management

As an example, you can write a token that can approve a zkApp account to send tokens from itself to another zkApp.

To implement this, you would have 3 contracts:

1. `MyToken`: The token contract. On instantiation, serves as the zkAppManagerContract for all instances of the token.

> 2. `TokenUser`: A zkApp contract which will manage tokens. This one is instantiated on the Mina (default) tokenId

Reviewer: ```suggestion
2. `TokenUser`: A zkApp contract that manages tokens. This one is instantiated on the Mina (default) tokenId.
```

> 3. `TokenHolder`: A zkApp contract which will store tokens on behalf of the `TokenUser` contract. This one is instantiated on the `MyToken` instance TokenId.
> 
> You can find the full code [here](https://github.com/o1-labs/docs2/tree/main/examples/zkapps/11-advanced-account-updates/).

Reviewer: ```suggestion
3. `TokenHolder`: A zkApp contract that stores tokens on behalf of the `TokenUser` contract. This one is instantiated on the `MyToken` instance TokenId.

The full code is provided in the [examples/zkapps/11-advanced-account-updates/](https://github.com/o1-labs/docs2/tree/main/examples/zkapps/11-advanced-account-updates/) file.
```

> A token standard library is in the above which will abstract much of these details away too for standard use. The illustration here though should help explain what that library is doing and how to think about zkAppManagerAccounts in detail.
> 
> These contracts can be setup as follows:

Reviewer: ```suggestion
As shown in the example file, a token standard library abstracts many of these details away for standard use. The illustration here helps explain what that library is doing and how to think about zkAppManagerAccounts in detail.

These contracts can be set up as follows:
```

```typescript
  const myTokenInstance = new MyToken(myTokenAddr);
  const tokenUserInstance = new TokenUser(tokenUserAddr);
  const tokenHolderInstance = new TokenHolder(tokenUserAddr, myTokenInstance.token.id);
```

And deployed as follows:

```typescript
  const deploy_txn = await Mina.transaction(deployerAccount, () => {
    let feePayerUpdate = AccountUpdate.fundNewAccount(deployerAccount, 4);
    feePayerUpdate.send({ to: myTokenAddr, amount: accountFee });

    myTokenInstance.deploy();
    tokenUserInstance.deploy();
    tokenHolderInstance.deploy()

    myTokenInstance.approveDeploy(tokenHolderInstance.self);
  });
```

> The special thing to note here is the `approveDeploy(...)` on the `myTokenInstance`. Because `tokenHolderInstance` is going to exist as a token account, it needs approval to be deployed.

Reviewer: ```suggestion
The special thing to note here is the `approveDeploy(...)` on the `myTokenInstance()` method. Because `tokenHolderInstance()` is going to exist as a token account, it needs approval to be deployed.
```

Reviewer: is `myTokenInstance()` a method? If so, let's say so (or omit the "the")

The `myTokenInstance.approveDeploy(...)` code is as follows:

> ```typescript
>   @method approveDeploy(deployUpdate: AccountUpdate) {
>     // TODO should anything else be asserted here for a "safe" deployment?
>     this.approve(deployUpdate, AccountUpdate.Layout.NoChildren);

Reviewer 2: ```suggestion
```typescript
  @method approveDeploy(deployUpdate: AccountUpdate) {
    this.approve(deployUpdate, AccountUpdate.Layout.NoChildren);
```
I think you should delete the TODO comment :D

The question of whether an account update is "safe" depends on the application.

For example, a manager contract for NFTs could store NFTs in the 8-field state of each user account. In that case, being safe means that users cannot arbitrarily change their 8-field state. So, this method here would check that the account update doesn't modify state.

In the case of a simple token contract, the only aspect it wants to control about user's accounts is the token balance. So it needs to check that the token balance behaves according to its rules. Since your methods does that, it's safe.

Author: Maybe let's say, "// NOTE you should include any other application-specific constraints you would like here, for example asserting that an account cannot arbitrarily change its 8-field state"?

Reviewer 2: Good idea, I'll include something to that effect in the code and maybe also in the tutorial text

    // see if balance change is zero
    let balanceChange = Int64.fromObject(deployUpdate.body.balanceChange);
    balanceChange.assertEquals(Int64.from(0));
  }
```

> This checks that the deployUpdate is a single account update, with no children - and that it isn't creating any additional tokens, and is starting out with a balance of zero.
> 
> Next, here is an example of sending tokens from the `tokenUser`/`tokenHolder` to another account:

Reviewer: ```suggestion
This checks that the deployUpdate is a single account update, with no children, that it isn't creating any additional tokens, and that it is starting out with a balance of zero.

Next, here is an example of sending tokens from the `tokenUser`/`tokenHolder` to another account:
```

```typescript
  const txn2 = await Mina.transaction(deployerAccount, () => {
    let feePayerUpdate = AccountUpdate.fundNewAccount(deployerAccount, 1);
    tokenUserInstance.sendMyTokens(UInt64.from(100), deployerAccount.toPublicKey());
  });
```

> The `fundNewAccount` call is included as the deployerAccount in this case doesn't have an existing token account.

Reviewer: ```suggestion
The `fundNewAccount` call is included as the deployerAccount,  in this case it doesn't have an existing token account.
```

`tokenUserInstance.sendMyTokens(...)` is implemented as follows:

```typescript
  @method sendMyTokens(
    amount: UInt64, 
    destination: PublicKey
  ) {
    const tokenHolder = this.tokenHolder;
    tokenHolder.transfer(destination, amount);
    this.tokenContract.approveTransfer(tokenHolder.self);
  }
```

Which calls the `transfer` contract on `tokenHolder`:

```typescript
  @method transfer(to: PublicKey, amount: UInt64) {
    const toAccountUpdate = AccountUpdate.create(to, this.tokenId);
    toAccountUpdate.balance.addInPlace(amount);
    toAccountUpdate.body.mayUseToken =
      AccountUpdate.MayUseToken.InheritFromParent;

    this.balance.subInPlace(amount);

    this.self.body.mayUseToken = AccountUpdate.MayUseToken.ParentsOwnToken;
  }
```

And is approved by the token contract's `approveTransfer`:

```typescript
  @method approveTransfer(
    transferUpdate: AccountUpdate
  ) {
    // TODO should anything else be asserted here for a "safe" deployment?

>     this.approve(transferUpdate);

Reviewer 2: the default for `this.approve()` is to require the child to have no children. so this shouldn't work when passed the `transfer` account update, which creates a child as the receiver. The only pattern that works and is safe to my knowledge is the "approveUpdateAndSend" thing in the WrappedMina example

Reviewer 2: yes 👍🏻

    // TODO should this also assert that all accountUpdates / balanceChanges are for the correct tokenId?
    this.assertHasNoBalanceChange(transferUpdate);
  }
```

> This uses a helper `assertHasNoBalanceChange` which walks the accountUpdates being approved and ensures that the balance changes haven't created or destroyed any tokens.
> 
> The account updates for this transcation looks as follows:

Reviewer: ```suggestion
This uses a helper `assertHasNoBalanceChange` that walks the accountUpdates being approved and ensures that the balance changes haven't created or destroyed any tokens.

The account updates for this transaction look like:
```

Reviewer: oh, interesting, what do we mean by "that walks"?
(manages? or observes? or?)

<img src="/img/tutorial-11-txn2.png" width="95%" />

> This shows both the account update for the fee payment - and the main chain which does the token operations. You can see the call to `tokenUser` at top, followed by a call to `myToken`, that is doing the approving. This approves two accountUpdates - one of which is taking tokens away from the `tokenUser`, and has the `mayUseToken` field on - and the other of which is sending tokens to the deployer account.
> 
> Because of the assertions in the `myToken` approval account, the balances must match out between these two account updates. Which it does, and so a valid `proof` can be constructed.
> 
> As noted above, a token standard library is in the above which will abstract much of these details away too for standard use.

Reviewer: ```suggestion
This shows both the account update for the fee payment and the main chain that does the token operations. You can see the call to `tokenUser` at the top, followed by a call to `myToken` that is doing the approving. This approves two accountUpdates: one update takes tokens away from the `tokenUser` and has the `mayUseToken` field on, and the other update sends tokens to the deployer account.

Because of the assertions in the `myToken` approval account, the balances must match between these two account updates. Which it does, and so a valid `proof` can be constructed.

As noted earlier, a token standard library abstracts many of these details away for standard use.
```

## Conclusion

Congratulations! You have explored some of the advanced features of AccountUpdates. You can now build with Token zkAppManagers.

> To see how to apply some of this in a more complex example, see the "wrappedMinaToken" implementation [here](https://github.com/o1-labs/docs2/tree/main/examples/zkapps/11-advanced-account-updates/src/wrappedMinaMain.ts)
> 
> and learned about visualizing the AccountUpdates for a set of transactions. You can build more complicated transactions that involve multiple zkApps. This tutorial builds a foundational understanding of how SnarkyJS and zkApps work to enable permissions, preconditions, and composability.

Reviewer: ```suggestion
To apply account updates in a more complex example, see the [wrappedMinaToken](https://github.com/o1-labs/docs2/tree/main/examples/zkapps/11-advanced-account-updates/src/wrappedMinaMain.ts) implementation example.
```
